src/qumin/representations/patterns.py

```diff
@@ -706,143 +831,295 @@ def format_as_features(left, right):
+        # return table
+
+    def from_csv(self, path, patterns_map, collection,
+                 paradigms, defective=True, overabundant=True):
+        """
+        Read a patterns dataframe for a specific pair of cells
+
+        Arguments:
+            paradigms (pandas.DataFrame): a paradigms dataframe, with form id's as index.
+            defective (bool): whether to consider defective lexemes.
+            overabundant (bool): whether to consider overabundance.
+            collection (defaultdict): a defaultdict to avoid recomputing
+                patterns from strings.
+        """
+
+        def read_pattern(string):
+            """
+            Reads patterns from string representations.
+            Arguments:
+                string (str): a string representation of a pattern.
+            """
+            if string and not pd.isnull(string):
+                if string in collection[cells]:
+                    result = collection[cells][string]
                 else:
-                    done = []
-                    log.debug("All alignments of {}, {}".format(a, b))
-                    for aligned in alignment.align_auto(a.tokens, b.tokens, insert_cost, sub_cost):
-                        log.debug((aligned))
-                        new_rule = Pattern(cells, aligned, aligned=True)
-                        new_rule.lexemes = {(row.name, a, b)}
-                        log.debug("pattern: " + str(new_rule))
-                        if str(new_rule) not in done:
-                            done.append(str(new_rule))
-                            if new_rule._gen_alt:
-                                alt = new_rule.to_alt(exhaustive_blanks=False, use_gen=True)
-                                log.debug("gen alt: " + str(alt))
-                            else:
-                                alt = new_rule.to_alt(exhaustive_blanks=False)
-                            t = _get_pattern_matchtype(new_rule, cells[0], cells[1])
-                            collection[alt][t].append(new_rule)
-
-    def attribute_best_pattern(sorted_collection, l, a, b):
-        for p in sorted_collection:
-            if (l, a, b) in p.lexemes:
-                return p
-
-    def _score(p, cells, forms, index):
-        def test_all(row, p, cells):
-            correct = 0
-            l, a_list, b_list = row
-            if a_list == b_list:  # If the lists are identical, do not compute product
-                pairs = [(x, x) for x in b_list]
+                    pattern = Pattern._from_str(cells, string)
+                    collection[cells][string] = pattern
+                    result = pattern
+                return result
+            if defective:
+                return None
             else:
-                pairs = product(a_list, b_list)
-            for a, b in pairs:
+                return np.nan
+
+        table = pd.read_csv(path, sep=",", dtype="str")
+
+        cells = tuple(path.name.split('.')[0].split('_')[1].split('-'))
```

Comment:
This is quite bad:
(1) It doesn't work (see eg file name: `phon_ind.prs.2.pl-sbjv.prs.2.pl.csv`) => this I can fix
(2) I really dislike having to parse file names: didn't we talk about having a neat json object associating cell pairs to file names ? That's a bigger thing to change.

Reply:
I guess you're right !